File: src/bot/commands/war/remaining.ts
import { AutocompleteInteraction, SlashCommandBuilder, type ChatInputCommandInteraction } from 'discord.js';
import { Command } from '../../struct/Command.js';
import { baseEmbed } from '../../../util/embeds.js';
import { EMOJIS } from '../../../util/emojis.js';
import { relativeTimestamp } from '../../../util/helper.js';

/** Members who still have attacks left in the current war. */
export default class RemainingCommand extends Command {
  public constructor() {
    super('remaining', { category: 'war', description: 'Show remaining war attacks.', defer: true });
  }

  public builder() {
    return new SlashCommandBuilder()
      .setName('remaining')
      .setDescription('Show remaining war attacks.')
      .addStringOption((opt) => opt.setName('tag').setDescription('Clan tag or alias.').setAutocomplete(true));
  }

  public async autocomplete(interaction: AutocompleteInteraction, args: Record<string, any>) {
    await interaction.respond(await this.client.resolver.clanAutocomplete(interaction.guildId, String(args.tag ?? '')));
  }

  public async exec(interaction: ChatInputCommandInteraction, args: Record<string, any>) {
    const { data: clan, error } = await this.client.resolver.resolveClan(interaction, args.tag);
    if (!clan) return interaction.editReply(error!);

    const war = await this.client.coc.getCurrentWar(clan.tag);
    if (!war || war.state === 'notInWar') return interaction.editReply('Clan is not currently in a war.');
    const w = war as any;
    const perMember = w.attacksPerMember ?? 2;

    const lines = w.clan.members
      .map((m: any) => ({ m, used: m.attacks?.length ?? 0 }))
      .filter((x: any) => x.used < perMember)
      .sort((a: any, b: any) => a.m.mapPosition - b.m.mapPosition)
      .map((x: any) => `${EMOJIS.SWORD} **${x.m.name}** — ${perMember - x.used} left`);

    const embed = baseEmbed(this.client, interaction)
      .setTitle(`${w.clan.name} — Remaining Attacks`)
      .setDescription(lines.length ? lines.join('\n').slice(0, 4000) : 'All attacks used! 🎉');

    if (w.state === 'inWar') embed.setFooter({ text: 'War ends' }).setTimestamp(w.endTime);
    else embed.addFields({ name: 'Battle day', value: relativeTimestamp(w.startTime) });

    return interaction.editReply({ embeds: [embed] });
  }
}

File: cogs/war/commands/remaining.py
"""/war remaining — members with attacks left."""

import coc
import discord

from utils.embeds import base_embed, error_embed
from utils.emojis import E_SWORD
from utils.helpers import discord_relative
from utils.resolver import resolve_clan


async def remaining(interaction: discord.Interaction, tag: str = None):
    await interaction.response.defer()
    try:
        clan = await resolve_clan(interaction, tag)
    except ValueError as err:
        await interaction.followup.send(embed=error_embed(str(err)))
        return

    try:
        war = await interaction.client.coc_client.get_current_war(clan.tag)
    except coc.PrivateWarLog:
        await interaction.followup.send(embed=error_embed("This clan's war log is private."))
        return
    except Exception:
        await interaction.followup.send(embed=error_embed("Clash of Clans API request failed, try again in a bit."))
        return

    if war is None or war.state == "notInWar":
        await interaction.followup.send(embed=error_embed(f"{clan.name} is not currently in a war."))
        return

    per_member = war.attacks_per_member or 2
    laggards = []
    for m in sorted(war.clan.members, key=lambda x: x.map_position):
        used = len(m.attacks)
        if used < per_member:
            laggards.append(f"{E_SWORD} **{m.name}** — {per_member - used} left")

    embed = await base_embed(interaction, title=f"{war.clan.name} — Remaining Attacks")
    embed.description = "\n".join(laggards)[:4000] if laggards else "All attacks used! 🎉"
    if war.state == "preparation":
        embed.add_field(name="Battle day", value=discord_relative(war.start_time.time), inline=False)
    else:
        embed.add_field(name="War ends", value=discord_relative(war.end_time.time), inline=False)

    await interaction.followup.send(embed=embed)
